proposals/manual-signature-hsm-design-proposal.md

```diff
@@ -0,0 +1,159 @@
+
+- Extract the canonical `bodyBytes` that require signing.
+- Sign externally (e.g., with an HSM).
+- Inject the signature back into the transaction using supported SDK methods.
+
+Currently, in the JavaScript SDK, `PrivateKey.sign(transaction.toBytes())` does not produce a network-valid signature because it signs the entire transaction, not the canonical `bodyBytes`.
+
+This proposal introduces a single new public getter to expose these canonical bytes: `signableBodyBytesList`.
+
+## New API
+
+### `Transaction.signableBodyBytesList`
+
+**Type**: `Uint8Array[]`  
+**Description**: Returns an array of `bodyBytes` for each node that the transaction targets. These are the exact serialized protobuf `TransactionBody` values used to compute the transaction hash and must be signed externally.
+
+The transaction **must be frozen** before calling this method. If not frozen, the SDK will throw an error to ensure the returned bytes are stable.
+
+This API is intended for secure signing workflows via HSMs, KMS, or offline signing mechanisms.
+
+- Expose `bodyBytes` through the new `signableBodyBytesList` getter.
+- Ensure this getter throws if the transaction is not frozen (`_requireFrozen()`).
+- Ensure ordering of returned `bodyBytes` aligns with the order of `nodeAccountIds`.
+
+**POC Implementation**:
+
+```javascript
+    get signableBodyBytesList() {
+        this._requireFrozen();
+
+        const result = [];
+        const nodeAccountIds = this._nodeAccountIds.list;
+
+        for (let i = 0; i < nodeAccountIds.length; i++) {
+            const nodeId = nodeAccountIds[i];
+            const signedTransaction = this._signedTransactions.get(i);
+
+            if (!signedTransaction?.bodyBytes) {
+                throw new Error(
+                    'Missing bodyBytes for node ' + nodeId.toString()
+                );
+            }
+
+            result.push(signedTransaction.bodyBytes);
+        }
+
+        return result;
+    }
+```
+
+**JavaScript Example**:
+
+    ```javascript
+    const transactionSignableBodyBytesList = transaction.signableBodyBytesList;
+    // send each transactionSignableBodyBytesList[i] to HSM for signing
+    ```
+
+---
+
+## Test Plan
+
+1. **Given** a transaction with one node,  
+   **When** its `bodyBytes` is signed externally and injected,  
+   **Then** the transaction executes successfully.
+
+2. **Given** a transaction with multiple nodes,  
+   **When** all node-specific signatures are correctly applied,  
+   **Then** the transaction executes successfully with retries allowed.
+
+3. **Given** a transaction with an invalid or mismatched signature applied via `signableBodyBytesList`,  
+   **Then** the transaction fails with `INVALID_SIGNATURE`.
+
+4. **Given** a call to `signableBodyBytesList` on a frozen transaction,  
+   **Then** the returned array length equals the number of node IDs.
+
+5. **Given** a call to `signableBodyBytesList` before freezing,  
+   **Then** an error is thrown.
+
```

Comment:
6. Check if chunked transaction with multi-node will return the proper ordering as the one you said in the proposal. 

Not sure if the ordering of the signed bodyBytes array will be:

```
chunk1 node1
chunk1 node2
chunk1 node3
chunk2 node1
chunk2 node2
chunk2 node3
```

or 
```
- chunk1 node1
- chunk2 node1
- chunk1 node2
- chunk2 node2
 ```